relentless/potential/pair.py

```diff
@@ -156,6 +155,298 @@ def _derivative(self, param, r, epsilon, sigma, **params):
+        super().__init__(types=types,params=params)
+
+    def from_array(self, pair, r, u):
+        """Sets up the potential from arrays of knot positions and energies.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to set up the potential.
+        r : array_like
+            The array of knot positions.
+        u : array_like
+            The array of energies at each knot position.
+
+        Raises
+        ------
+        ValueError
+            If the array of r values does not have the same length as the number of knots.
+        ValueError
+            If the array of u values does not have the same length as the number of knots.
+
+        """
+        # check that r and u have the right shape
+        if len(r) != self.num_knots:
+            raise ValueError('r must have the same length as the number of knots')
+        if len(u) != self.num_knots:
+            raise ValueError('u must have the same length as the number of knots')
+
+        # convert to r,knot form given the mode
+        rs = np.asarray(r, dtype=np.float64)
+        ks = np.asarray(u, dtype=np.float64)
+        if self.mode == 'diff':
+            # difference is next knot minus my knot, with last knot fixed at its current value
+            ks[:-1] -= ks[1:]
+
+        # make all knots variables, but hold all r and the last knot const
+        for i in range(self.num_knots):
+            ri,ki = self._knot_params(i)
+            if self.coeff[pair][ri] is None:
+                self.coeff[pair][ri] = core.Variable(rs[i],const=True)
+            else:
+                self.coeff[pair][ri].value = rs[i]
+            if self.coeff[pair][ki] is None:
+                self.coeff[pair][ki] = core.Variable(ks[i],const=(i==self.num_knots-1))
+            else:
+                self.coeff[pair][ki].value = ks[i]
+
+    def _knot_params(self, i):
+        """Returns the knot variables.
+
+        Parameters
+        ----------
+        i : int
+            Key for the knot variable.
+
+        Returns
+        -------
+        str
+            The keyed r value.
+        str
+            The keyed knot value.
+
+        Raises
+        ------
+        ValueError
+            If the knot key is not an integer.
+
+        """
+        if not isinstance(i, int):
+            raise ValueError('Knots are keyed by integers')
+        return 'r-{}'.format(i), 'knot-{}'.format(i)
+
+    def _energy(self, r, params, **pars):
+        """Evaluates the spline potential energy.
+
+        Parameters
+        ----------
+        r : array_like
+            The value or values of r at which to evaluate the energy.
+        params : dict
+            The parameters (r and knot values) for the potential function.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the energy evaluated at r.
+
+        """
+        r,u,s = self._zeros(r)
+        u = self._interpolate(params)(r)
```

Comment:
`_interpolate` is using `params` as a dict, which is why it doesn't need to be packed again. If you wanted to make the signatures the same, you could change this to `**params` and then modify it below to accept `**params` as well.